text/0000-iri-v181-transaction-and-bundle/0000-iri-v181-transaction-and-bundle.md

```diff
@@ -0,0 +1,980 @@
+| `tag`                             | arbitrary user-defined value                            | 81              | `add_message` or `add_input` or `add_output`  |
+| `attachment_timestamp`            | the timestamp for when Proof-of-Work is completed       | 27              | `calculate_proof_of_work`                     |
+| `attachment_timestamp_lowerbound` | *not specified*                                         | 27              | `calculate_proof_of_work`                     |
+| `attachment_timestamp_upperbound` | *not specified*                                         | 27              | `calculate_proof_of_work`                     |
+| `nonce`                           | the Proof-of-Work nonce of the transaction              | 81              | `calculate_proof_of_work`                     |
+
+Each transaction is uniquely identified by its *transaction hash*, which is calculated based on all fields of the
+transaction. Note that the transaction hash is not part of the transaction.
+
+Because modifying the fields of a transaction would invalidate its hash, `Transaction` is immutable after creation.
+
+### `Transaction` struct
+
+A transaction is represented by the `Transaction` struct and follows the structure presented in the table above:
+
+```rust
+pub struct Transaction {
+    signature_or_message_fragment: SignatureOrMessageFragment,
+    address: Address,
+    value: Value,
+    obsolete_tag: Tag,
+    timestamp: Timestamp,
+    current_index: Index,
+    last_index: Index,
+    bundle: BundleHash,
+    trunk: TransactionHash,
+    branch: TransactionHash,
+    tag: Tag,
+    attachment_timestamp: Timestamp,
+    attachment_timestamp_lower_bound: Timestamp,
+    attachment_timestamp_upper_bound: Timestamp,
+    nonce: Nonce
+}
+```
+
+The `Transaction` type only contains two constructor methods getter methods for read-only borrows of its fields.
+
+```rust
+impl Transaction {
+    /// Create a `Transaction` from a reader.
+    pub fn from_reader<R: std::io::Read>(reader: R) -> Result<Self, TransactionError> {
+        unimplemented!()
+    }
+
+    /// Create a `Transaction` from a slice of bytes
+    pub fn from_slice(stream: &[u8]) -> Result<Self, TransactionError> {
+        unimplemented!()
+    }
+
+    pub fn signature_or_message_fragment(&self) -> &SignatureOrMessageFragment {
+        &self.signature_fragments
+    }
+
+    pub fn address(&self) -> &Address {
+        &self.address
+    }
+
+    pub fn value(&self) -> &Value {
+        &self.value
+    }
+}
+```
+
+We treat the types of the `Transaction` struct's fields as opaque newtypes for now so that we can flesh them out during
+the implementation phase or in future RFCs. Because this RFC implements the transaction format as of iri v1.8.1, the
+newtypes shall be constructed from byte slices of a certain length matching that of the reference implementation.
```

Comment:
a Java byte is an 'i8' in Rust, not an 'u8'. So if we want to match the reference implementation we should use 'i8' in this RFC as the proper equivalent. It also simplifies calculations as we tell the compiler about how we expect the bits in the byte to be interpreted (0xFF as -1, not as 255).

Reply:
I have added a sentence about converting from `&[i8]`. In general, I'd like to keep the representation as bytes, but conversion form `&[i8]` is fine. Also, we can change this down the road if we realize that this is not nice.